lib/ain-evm/src/backend.rs

```diff
@@ -38,11 +42,86 @@ pub struct Vicinity {
     pub block_randomness: Option<H256>,
 }
 
+#[derive(Debug, Clone)]
+struct OverlayData {
+    account: Account,
+    code: Option<Vec<u8>>,
+    storage: HashMap<H256, H256>,
+}
+
+#[derive(Debug, Clone)]
+struct Overlay {
+    state: HashMap<H160, OverlayData>,
+    changeset: Vec<HashMap<H160, OverlayData>>,
+    deletes: HashSet<H160>,
+}
+
+impl Overlay {
+    pub fn new() -> Self {
+        Self {
+            state: HashMap::new(),
+            changeset: Vec::new(),
+            deletes: HashSet::new(),
+        }
+    }
+
+    pub fn apply(
+        &mut self,
+        address: H160,
+        account: Account,
+        code: Option<Vec<u8>>,
+        mut storage: HashMap<H256, H256>,
+        reset_storage: bool,
+    ) {
+        if !reset_storage {
+            if let Some(existing_storage) = self.storage(&address) {
+                for (k, v) in existing_storage {
+                    storage.entry(*k).or_insert_with(|| *v);
+                }
+            }
```

Comment:
Not sure if I am seeing this correctly, but it seems like by doing this, we could potentially overwrite the current storage state with an old value with the same key in the existing storage?

Reply:
Yes, in the case where we would write multiple time to the same storage slot in a block (such as TD total_supply), we would overwrite the current storage with the updated value

Reply:
Isn't existing_storage the old value? Meaning that we are overwriting the updated value with the old value?

Reply:
So it keeps the `storage` entry as is with `.entry(*k)` or if the key does not exist in storage, insert the existing key val from existing_storage. Feels correct but we can add some test if needed

Reply:
Logic could probably be reversed since we can assume that `existing_storage` will grow and be on average bigger than `storage`